hydrolib/profile_optimizer/profile_optimizer/preprocessing.py

```diff
@@ -0,0 +1,173 @@
+
+    #Clip deze profiles geometrieën op de opgegeven optimization area
+    points_clip = gpd.clip(gdf_cross_loc, shapefile_area)
+    points_clip.to_file(str(output_folder/'cross_loc_geometry_within_optimization_area.shp'), driver = 'ESRI Shapefile')
+    points_clip.rename(columns={'cross_loc_geom': 'geometry'}, inplace=True)
+    points_clip = points_clip[['branchid', 'definitionid', 'geometry']]
+    return points_clip
+    
+
+def search_window(b_start_value, bandwidth_perc, iterations):
+    max_bound_b = b_start_value * (1 + (bandwidth_perc / 100))
+    min_bound_b = b_start_value * (1 - (bandwidth_perc / 100))
+    b_waardes_binnen_zoekruimte = np.linspace(min_bound_b, max_bound_b, iterations)
+    return b_waardes_binnen_zoekruimte
+
+# Ideas:
+# Checks with material (sand/clay/peat/etc) if talud & velocity are okay (vademecum)
+
+def determine_v_with_manning(d, talud, b, slope, kmanning):
+    A = ((talud * d) * d)+ (b * d)
+    R = A / (b + math.sqrt(d ** 2 + (d * talud) ** 2) * 2)
+    V = R ** (2 / 3) * kmanning * slope ** (1 / 2)
+    return V
+
+def check_QVA(Q_target, d, talud, b, slope, kmanning, allowed_variation=0.05):
+    # Belangrijke keuze: 0.05% aanpassen (stop dit in notitie)
+    if b < 0:
+        b = 1
+        print("First guess for bottom width was negative. Trying to find a positive bottom width, starting at 1 m...")
+
+    V = determine_v_with_manning(d, talud, b, slope, kmanning)
+    A = ((talud * d) * d)+ (b * d)
+    Q = V * A
+    print(f"Initial: width: {b:.2f}, V: {V:.4f}, Q: {Q:.4f}")
+    deviation_from_target = (Q - Q_target) / Q_target
+    counter = 0
+    while abs(deviation_from_target) > allowed_variation:
+        counter += 1
+        stepsize = 0.05  # %
+        if deviation_from_target < - allowed_variation:
+            b *= (1+stepsize)
+            V = determine_v_with_manning(d, talud, b, slope, kmanning)
+            A = ((talud * d) * d) + (b * d)
+            Q = V * A
+            deviation_from_target = (Q - Q_target) / Q_target
+            print(f"Adjustment {counter}: new width: {b:.2f}, V: {V:.4f}, Q: {Q:.4f}")
+        elif deviation_from_target > allowed_variation:
+            b *= (1-stepsize)
+            V = determine_v_with_manning(d, talud, b, slope, kmanning)
+            A = ((talud * d) * d) + (b * d)
+            Q = V * A
+            deviation_from_target = (Q - Q_target) / Q_target
+            print(f"Adjustment {counter}: new width: {b:.2f}, V: {V:.4f}, Q: {Q:.4f}")
+        if counter == 30:
+            print("Failed to find suitable initial bottom width in 30 tries, please check if your inputs are correct "
+                  "and use the returned bottom width with caution.")
+            return b
+    else:
+        return b
+
+
+def bottom_width(kmanning, slope, talud, diepte, V_target):
+    # Let op: dit is symmetrisch!
+    R_23 = V_target / (kmanning * slope ** (1/2))
+    b = Symbol('b')
+    A = (b * diepte) + (diepte * diepte * talud)
```

Comment:
This formula is used multiple times, does it make sense to extract it as a separate function?

Reply:
Hah, smart :)